query/sources/impl/src/main/java/org/codice/ddf/admin/sources/commons/utils/WfsSourceUtils.java

```diff
@@ -0,0 +1,161 @@
+import static org.codice.ddf.admin.sources.services.WfsServiceProperties.resolveWfsFactoryPid;
+
+import java.util.List;
+import java.util.Map;
+
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+
+import org.codice.ddf.admin.api.action.Message;
+import org.codice.ddf.admin.common.Result;
+import org.codice.ddf.admin.common.fields.common.AddressField;
+import org.codice.ddf.admin.common.fields.common.CredentialsField;
+import org.codice.ddf.admin.common.fields.common.UrlField;
+import org.codice.ddf.admin.sources.fields.type.SourceConfigUnionField;
+import org.codice.ddf.admin.sources.fields.type.WfsSourceConfigurationField;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.Document;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+public class WfsSourceUtils {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(WfsSourceUtils.class);
+
+    public static final Map<String, String> GET_CAPABILITIES_PARAMS = ImmutableMap.of("service",
+            "WFS",
+            "request",
+            "GetCapabilities",
+            "AcceptVersions",
+            "2.0.0,1.0.0");
+
+    private static final List<String> URL_FORMATS = ImmutableList.of("https://%s:%d/services/wfs",
+            "https://%s:%d/wfs",
+            "http://%s:%d/services/wfs",
+            "http://%s:%d/wfs");
+
+    private static final String WFS_VERSION_EXP = "/wfs:WFS_Capabilities/attribute::version";
+
+    private final RequestUtils requestUtils;
+
+    public WfsSourceUtils() {
+        this(new RequestUtils());
+    }
+
+    public WfsSourceUtils(RequestUtils requestUtils) {
+        this.requestUtils = requestUtils;
+    }
+
+    /**
+     * Attempts to verify the given URL as a functional WFS endpoint
+     *
+     * @param urlField URL to probe for WFS capabilities
+     * @param creds    optional username and password to add to Basic Auth header
+     * @return a {@link Result} containing the {@link UrlField} or an {@link org.codice.ddf.admin.common.message.ErrorMessage} on failure.
+     */
+    public List<Message> sendWfsCapabilitiesRequest(UrlField urlField, CredentialsField creds) {
+        Result<String> result = requestUtils.sendGetRequest(urlField, creds, GET_CAPABILITIES_PARAMS);
+        return result.allMessages();
+    }
+
+    /**
+     * Attempts to discover a WFS endpoint at a given hostname and port
+     *
+     * @param addressField address to probe for WFS capabilities
+     * @param creds        optional username to add to Basic Auth header
+     * @return @return a {@link Result} containing the {@link UrlField} or an {@link org.codice.ddf.admin.common.message.ErrorMessage} on failure.
+     */
+    public Result<UrlField> discoverWfsUrl(AddressField addressField, CredentialsField creds) {
+        Result<UrlField> defaultResult = new Result<>();
```

Comment:
nit, probably better to only create this object when it is needed in the `orElse` section of the stream below